src/librbd/migration/QCOWFormat.cc

```diff
@@ -0,0 +1,1392 @@
+    while (image_length > 0) {
+      auto intra_cluster_offset = image_offset & (cluster_size - 1);
+      auto intra_cluster_length = cluster_size - intra_cluster_offset;
+      auto cluster_length = std::min(image_length, intra_cluster_length);
+
+      ldout(cct, 20) << "image_offset=" << image_offset << ", "
+                     << "image_length=" << image_length << ", "
+                     << "cluster_length=" << cluster_length << dendl;
+
+
+      cluster_extents->emplace_back(0, cluster_length, intra_cluster_offset,
+                                   image_offset, buffer_offset);
+
+      image_offset += cluster_length;
+      image_length -= cluster_length;
+      buffer_offset += cluster_length;
+    }
+  }
+}
+
+} // anonymous namespace
+
+template <typename I>
+struct QCOWFormat<I>::Cluster {
+  const uint64_t cluster_offset;
+  bufferlist cluster_data_bl;
+
+  Cluster(uint64_t cluster_offset) : cluster_offset(cluster_offset) {
+  }
+};
+
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::migration::QCOWFormat::ClusterCache: " \
+                           << this << " " << __func__ << ": "
+
+template <typename I>
+class QCOWFormat<I>::ClusterCache {
+public:
+  ClusterCache(QCOWFormat* qcow_format)
+    : qcow_format(qcow_format),
+      m_strand(*qcow_format->m_image_ctx->asio_engine) {
+  }
+
+  void get_cluster(uint64_t cluster_offset, uint64_t cluster_length,
+                   uint64_t intra_cluster_offset, bufferlist* bl,
+                   Context* on_finish) {
+    auto cct = qcow_format->m_image_ctx->cct;
+    ldout(cct, 20) << "cluster_offset=" << cluster_offset << dendl;
+
+    // cache state machine runs in a single strand thread
+    boost::asio::dispatch(
+      m_strand,
+      [this, cluster_offset, cluster_length, intra_cluster_offset, bl,
+       on_finish]() {
+        execute_get_cluster(cluster_offset, cluster_length,
+                            intra_cluster_offset, bl, on_finish);
+      });
+  }
+
+private:
+  typedef std::tuple<uint64_t, uint64_t, bufferlist*, Context*> Completion;
+  typedef std::list<Completion> Completions;
+
+  QCOWFormat* qcow_format;
+  boost::asio::io_context::strand m_strand;
+
+  std::shared_ptr<Cluster> cluster;
+  std::unordered_map<uint64_t, Completions> cluster_completions;
+
+  void execute_get_cluster(uint64_t cluster_offset, uint64_t cluster_length,
+                           uint64_t intra_cluster_offset, bufferlist* bl,
+                           Context* on_finish) {
+    auto cct = qcow_format->m_image_ctx->cct;
+    ldout(cct, 20) << "cluster_offset=" << cluster_offset << dendl;
+
+    if (cluster && cluster->cluster_offset == cluster_offset) {
+      // most-recent cluster matches
+      bl->substr_of(cluster->cluster_data_bl, intra_cluster_offset,
+                    cluster_length);
+      boost::asio::post(*qcow_format->m_image_ctx->asio_engine,
+                        [on_finish]() { on_finish->complete(0); });
```

Comment:
it looks like this line is causing build failures in fedora on a pacific snapshot:

https://kojipkgs.fedoraproject.org//work/tasks/1687/60471687/build.log
```
In file included from /usr/include/boost/asio/post.hpp:126,
                 from /usr/include/boost/asio/basic_socket.hpp:28,
                 from /usr/include/boost/asio/basic_datagram_socket.hpp:20,
                 from /usr/include/boost/asio.hpp:24,
                 from /builddir/build/BUILD/ceph-16.0.0/src/common/error_code.h:22,
                 from /builddir/build/BUILD/ceph-16.0.0/src/include/denc.h:50,
                 from /builddir/build/BUILD/ceph-16.0.0/src/include/encoding.h:41,
                 from /builddir/build/BUILD/ceph-16.0.0/src/include/uuid.h:9,
                 from /builddir/build/BUILD/ceph-16.0.0/src/include/types.h:21,
                 from /builddir/build/BUILD/ceph-16.0.0/src/librbd/Types.h:7,
                 from /builddir/build/BUILD/ceph-16.0.0/src/librbd/migration/QCOWFormat.h:8,
                 from /builddir/build/BUILD/ceph-16.0.0/src/librbd/migration/QCOWFormat.cc:4:
/usr/include/boost/asio/impl/post.hpp: In instantiation of 'auto boost::asio::post(ExecutionContext&, CompletionToken&&, typename std::enable_if<std::is_convertible<ExecutionContext&, boost::asio::execution_context&>::value>::type*) [with ExecutionContext = librbd::AsioEngine; CompletionToken = librbd::migration::QCOWFormat<ImageCtxT>::ClusterCache::execute_get_cluster<librbd::ImageCtx>::<lambda()>; typename std::enable_if<std::is_convertible<ExecutionContext&, boost::asio::execution_context&>::value>::type = void]':
/builddir/build/BUILD/ceph-16.0.0/src/librbd/migration/QCOWFormat.cc:148:24:   required from 'void librbd::migration::QCOWFormat<ImageCtxT>::ClusterCache::execute_get_cluster(uint64_t, uint64_t, uint64_t, ceph::bufferlist*, Context*) [with I = librbd::ImageCtx; uint64_t = long unsigned int; ceph::bufferlist = ceph::buffer::v15_2_0::list]'
/builddir/build/BUILD/ceph-16.0.0/src/librbd/migration/QCOWFormat.cc:1391:35:   required from here
/usr/include/boost/asio/impl/post.hpp:240:15: error: no type named 'executor_type' in 'struct librbd::AsioEngine'
  240 |       detail::initiate_post_with_executor<
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
  241 |         typename ExecutionContext::executor_type>(
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  242 |           ctx.get_executor()), token);
      |           ~~~~~~~~~~~~~~~~~~~
```
if `librbd::AsioEngine` is intended to be used as an ExecutionContext, it needs to expose an `executor_type`:
```diff
+  using executor_type = boost::asio::io_context::executor_type;
   inline boost::asio::io_context::executor_type get_executor() {
     return m_io_context.get_executor();
   }
```
the conversion operator `operator boost::asio::io_context&()` wouldn't be called in this case. maybe it's best to post directly to the io_context?
```c++
boost::asio::post(qcow_format->m_image_ctx->asio_engine->get_io_context(),
```

Reply:
Hmm, I use F32 and F33 for builds so I am interested in why I never saw this.

Reply:
... also, please open a tracker ticket if this needs to be fixed instead of commenting on an old, merged PR :wink:

Reply:
https://tracker.ceph.com/issues/48989

Reply:
> Hmm, I use F32 and F33 for builds so I am interested in why I never saw this.

this is on f34/rawhide, where they have just updated to boost-1.75.